### PR TITLE
Reduce asm size of CollectionsMarshal.AsSpan

### PR DESCRIPTION
**Pull request by stephentoub**
Using the span constructor brings with it some branches and code we don't need due to knowledge of `List<T>`'s implementation, in particular that a) it always has an array (even if empty), and b) its array is invariant.

Erroneous use of AsSpan while the list is being mutated could have also led to an ArgumentOutOfRangeException from the span's constructor, which doesn't make much sense in this context. I've changed it to be an InvalidOperationException about the invalid concurrent use.

| Method | Toolchain         | Code Size |
|------- |------------------ |----------:|
| AsSpan | \main\corerun.exe |      99 B |
| AsSpan | \pr\corerun.exe   |      58 B |
|        |                   |           |
| Sum    | \main\corerun.exe |     112 B |
| Sum    | \pr\corerun.exe   |      75 B |

```C#
using BenchmarkDotNet.Attributes;
using BenchmarkDotNet.Running;
using System.Runtime.InteropServices;

BenchmarkSwitcher.FromAssembly(typeof(Tests).Assembly).Run(args);

[DisassemblyDiagnoser]
public class Tests
{
    private readonly List<string> _list = new() { "a", "b", "c" };

    [Benchmark]
    public Span<string> AsSpan() => CollectionsMarshal.AsSpan(_list);

    [Benchmark]
    public int Sum()
    {
        int total = 0;
        foreach (string s in CollectionsMarshal.AsSpan(_list))
        {
            total += s.Length;
        }
        return total;
    }
}
```

### CODE REVIEW

**Comment by jkotas** [comment]
Why is the JIT not able to optimize the original code well?

cc @dotnet/jit-contrib

**Comment by stephentoub** [comment]
> Why is the JIT not able to optimize the original code well?

For starters I assume because it doesn't know that it can elide the null check on the array and elide the variance check on the array, since it's not privvy to the information about how the list was constructed.

**Comment by stephentoub** [comment]
> For the variance check

Even if that's fixed for string, though, I imagine the JIT would have a lot more trouble doing it for an arbitrary non-sealed reference type.

**Comment by EgorBo** [comment]
> > For the variance check
> 
> Even if that's fixed for string, though, I imagine the JIT would have a lot more trouble doing it for an arbitrary non-sealed reference type.

Sure, but looks like it's should be a general goodness to improve this in JIT too if possible

**Comment by stephentoub** [comment]
> Sure, but looks like it's should be a general goodness to improve this in JIT too if possible

Agreed. Just commenting that this change still improves upon what I imagine the JIT will be able to safely do, unless we build knowledge of `List<T>` into it directly.

**Comment by jkotas** [comment]
> this change still improves upon what I imagine the JIT will be able to safely

Yea, I agree. LGTM.

**Review by jkotas — APPROVED** [review]

**Comment by EgorBo** [comment]
> > > For the variance check
> > 
> > 
> > Even if that's fixed for string, though, I imagine the JIT would have a lot more trouble doing it for an arbitrary non-sealed reference type.
> 
> Sure, but looks like it's should be a general goodness to improve this in JIT too if possible

Handled via https://github.com/dotnet/runtime/pull/96794

File: src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/CollectionsMarshal.cs [hunk 1]
```diff
@@ -17,8 +18,28 @@ public static class CollectionsMarshal
...
-            => list is null ? default : new Span<T>(list._items, 0, list._size);
+        {
+            Span<T> span = default;
+            if (list is not null)
+            {
+                int size = list._size;
+                T[] items = list._items;
+                Debug.Assert(items is not null, "Implementation depends on List<T> always having an array.");
+
+                if ((uint)size >= (uint)items.Length)
```
Comment (stephentoub) [inline src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/CollectionsMarshal.cs:31]:
```suggestion
                if ((uint)size > (uint)items.Length)
```

File: src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/CollectionsMarshal.cs [hunk 2]
```diff
@@ -17,8 +18,28 @@ public static class CollectionsMarshal
...
+                Debug.Assert(items is not null, "Implementation depends on List<T> always having an array.");
+
+                if ((uint)size > (uint)items.Length)
+                {
+                    // List<T> was erroneously mutated concurrently with this call, leading to a count larger than its array.
+                    ThrowHelper.ThrowInvalidOperationException_ConcurrentOperationsNotSupported();
+                }
+
+                Debug.Assert(typeof(T[]) == list._items.GetType(), "Implementation depends on List<T> always using a T[] and not U[] where U : T.");
+                span = new(ref MemoryMarshal.GetArrayDataReference(items), size);
```
Comment (jkotas) [inline src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/CollectionsMarshal.cs:38]:
```suggestion
                span = new Span<T>(ref MemoryMarshal.GetArrayDataReference(items), size);
```
For compliance with https://github.com/dotnet/runtime/blob/main/docs/coding-guidelines/coding-style.md#c-coding-style

**Pull request merged**